### PR TITLE
Replacing multiple components on the product page with a single hook.

### LINKED ISSUE #37281: Refactor block editor product page logic using hook
As requested in [this comment](https://github.com/woocommerce/woocommerce/pull/37064#discussion_r1130713116):
>Instead of separating this logic into two separate (ultimately 3) components would it be worth just moving the product logic into a custom hook, that either returns a new auto draft product or an existing product?

>That way you can just call const product = useProductEntityRecord( productId ) and pass that straight into the Editor.
Only suggesting this, cause otherwise I think we are better off splitting this into separate files again (but we were actually trying to avoid that).

These changes will be made in the `product-page.tsx` file that is the router to the product block editor.

```[tasklist]
### Tasks
- [x] Replace components with hook
```

### PR DESCRIPTION
**Pull request by joelclimbsthings**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

Refactor `product-page.tsx` logic to utilize a single hook instead of three components, suggested as part of a [previous code review.](https://github.com/woocommerce/woocommerce/pull/37064#discussion_r1130713116) 

Closes #37281 .

### How to test the changes in this Pull Request:

1. Create a dummy product if you don't have one.
2. Enable the `block-editor-feature-enabled` feature flag.
3. Go to Products -> Add new.
4. You should have the block editor successfully load with the text `AUTO_DRAFT` present in the name field.
5. Get the preexisting productId and go to `/wp-admin/admin.php?page=wc-admin&path=/product/[PRODUCT_ID]`
6. The editor should successfully load with the name of the product present in the field.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		return getEditedEntityRecord( 'postType', 'product', productId );
+	} );
+
+	return product;
+};
+
+export function useProductEntityRecord(
+	productId: string | undefined
+): Product | undefined {
+	const selectedHook = useMemo( () => {
```
Comment (joelclimbsthings) [inline plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts]: Normally we don't want to allow the possibility of different hooks being fired in the same component on re-render, and this shouldn't allow that as only one would ever be fired on a particular screen. That said, if it feels uncomfortable we could just stick with something like the original approach using conditional components.

Comment (louwie17) [reply plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts]: I end up updating this, as React does have a specific rule to not call hooks conditionally: https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

This does work and is a nifty a trick around it, but I don't want to run into any edge cases.

File: packages/js/product-editor/src/components/block-editor/block-editor.tsx [hunk 2]
```diff
@@ -94,7 +94,7 @@ export function BlockEditor( {
 			synchronizeBlocksWithTemplate( [], _settings?.template ),
 			{}
 		);
-	}, [] );
+	}, [ product.id ] );
```
Comment (louwie17) [inline packages/js/product-editor/src/components/block-editor/block-editor.tsx:97]: Added this so it correctly syncs the template again if the product switches.

File: plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			  ( saveEntityRecord( 'postType', 'product', {
+					title: AUTO_DRAFT_NAME,
+					status: 'auto-draft',
+			  } ) as Promise< Product > );
+		getRecordPromise
+			.then( ( autoDraftProduct: Product ) => {
+				setProduct( autoDraftProduct );
+			} )
+			.catch( ( e ) => {
+				setProduct( undefined );
```
Comment (mdperez86) [inline plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts:36]: Maybe we can redirect to 404 page if the product was not found. And I would pipe the catch for the 404 use case only to the `getEntityRecord` function.

File: plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+				throw e;
```
Comment (mdperez86) [inline plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts:37]: If this line triggers the redirect then ignore the previous comment.

**Review by mdperez86 — APPROVED** [review]
LGTM! Nice job overall here @joelclimbsthings and @louwie17.

**Pull request merged**